consensus/src/rand/rand_gen/rand_store.rs

```diff
@@ -0,0 +1,227 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    experimental::buffer_manager::OrderedBlocks,
+    rand::rand_gen::{
+        block_queue::{BlockQueue, QueueItem},
+        types::{Proof, RandConfig, RandDecision, RandShare, Share},
+    },
+};
+use anyhow::ensure;
+use aptos_consensus_types::{
+    common::{Author, Round},
+    randomness::RandMetadata,
+};
+use aptos_logger::error;
+use std::collections::HashMap;
+
+struct ShareAggregator<S> {
+    shares: HashMap<Author, RandShare<S>>,
+    total_weight: u64,
+}
+
+impl<S: Share> ShareAggregator<S> {
+    fn new() -> Self {
+        Self {
+            shares: HashMap::new(),
+            total_weight: 0,
+        }
+    }
+
+    fn add_share(&mut self, weight: u64, share: RandShare<S>) {
+        self.shares.insert(*share.author(), share);
+        self.total_weight += weight;
+    }
+
+    fn try_aggregate<P: Proof<Share = S>>(
+        &self,
+        rand_config: &RandConfig,
+        rand_metadata: RandMetadata,
+    ) -> Option<RandDecision<P>> {
+        if self.total_weight < rand_config.threshold_weight() {
```

Comment:
should use rounded weight?

Reply:
yes, but I don't know how that works, so later